radiotherapyai/mask/test_convert.py

```diff
@@ -16,65 +16,133 @@
 
 """Testing the mask conversion to and from contour lines"""
 
+import pathlib
+from typing import TypedDict
+
+import matplotlib.pyplot as plt  # pyright: ignore [reportMissingTypeStubs, reportUnknownVariableType]
 import numpy as np
 
 from radiotherapyai.metrics import dice
 
-from .convert import contours_to_mask, mask_to_contours
+from .convert import Contours, Grid, contours_to_mask, mask_to_contours
+
+HERE = pathlib.Path(__file__).parent
+FIGURE_DIR = HERE / "test_figures"
 
 
-def test_mask_at_edge_of_image():
-    """Test the effect of a contour at the edge of an image"""
-    mask = np.array(
-        [
-            [255, 255, 127, 0],
-            [255, 255, 127, 0],
-            [127, 127, 64, 0],
-            [0, 0, 0, 0],
-        ],
-        dtype=np.uint8,
+def test_conversion_round_trip():
+    """Test a round trip of contours -> mask -> contours"""
+
+    cases: list[_TestCase] = []
+
+    cases.append(
+        {
+            "title": "edge-of-mask",
+            "x_grid": np.array([0, 1, 2, 3]),
+            "y_grid": np.array([10, 12, 14, 16]),
+            "contours": [np.array([(9, -0.5), (9, 2), (14, 2), (14, -0.5)])],
+            "dice_lower_bound": 0.90,
+        }
     )
 
-    y_grid = np.array([10, 12, 14, 16])
-    x_grid = np.array([0, 1, 2, 3])
+    t = np.linspace(0, 2 * np.pi)
+    x = 1.5 * np.sin(t)
+    y = np.cos(t) + 0.5
+    yx_coords = np.concatenate(  # pyright: ignore [reportUnknownMemberType]
+        [y[:, None], x[:, None]], axis=-1
+    )
 
-    ideal_expected_contour = np.array([(9, -0.5), (9, 2), (14, 2), (14, -0.5)])
-    ideal_contours = [ideal_expected_contour]
+    cases.append(
+        {
+            "title": "offset-ellipse",
+            "x_grid": np.linspace(-2, 2, 21),
+            "y_grid": np.linspace(-2, 2, 31),
+            "contours": [yx_coords],
+            "dice_lower_bound": 0.99,
+        }
+    )
 
-    # Marching squares doesn't perfectly follow the in-out proportion,
-    # instead, it cuts off corners.
-    marching_squares_expected_contour = np.array(
-        [(9, 0), (10, -0.5), (12, -0.5), (14, 0), (14, 1), (12, 2), (10, 2), (9, 1)]
+    cases.append(
+        {
+            "title": "right-angle-triangle",
+            "x_grid": np.linspace(0, 4, 5),
+            "y_grid": np.linspace(0, 10, 11),
+            "contours": [np.array([(0, 0), (10, 0), (0, 2), (0, 0)])],
+            # TODO: This results in an overlapping contour. Contour
+            # overlaps like this need to be cleaned up with shapely in
+            # post-processing.
+            "dice_lower_bound": 0.81,  # Sharp sub-pixel points are not handled well
```

Comment:
I'm working on having the network directly output the contour lines themselves using an approach inspired by:
https://arxiv.org/pdf/2202.01999.pdf

This will counter this "cut-off effect" which has relevance in organs like the optic nerves.